exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,262 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+)
+
+const (
+	prefixSplitter = "+"
+)
+
+type metricID *export.Descriptor
+
+// Exporter is an implementation of metric.Exporter that sends metrics to
+// Prometheus.
+type Exporter struct {
+	sync.RWMutex
+	registerer prometheus.Registerer
+	gatherer   prometheus.Gatherer
+
+	counters map[metricID]prometheus.Counter
+	gauges   map[metricID]prometheus.Gauge
+
+	onRegisterError func(e error)
+	counterVecs     map[string]*prometheus.CounterVec
+	gaugeVecs       map[string]*prometheus.GaugeVec
+}
+
+var _ export.Exporter = (*Exporter)(nil)
+
+// Options is a set of options for the tally reporter.
+type Options struct {
+	// Registerer is the prometheus registerer to register
+	// metrics with. Use nil to specify the default registerer.
+	Registerer prometheus.Registerer
+
+	// Gatherer is the prometheus gatherer to gather
+	// metrics with. Use nil to specify the default gatherer.
+	Gatherer prometheus.Gatherer
+
+	// DefaultHistogramBuckets is the default histogram buckets
+	// to use. Use nil to specify the default histogram buckets.
+	DefaultHistogramBuckets []float64
+
+	// DefaultSummaryObjectives is the default summary objectives
+	// to use. Use nil to specify the default summary objectives.
+	DefaultSummaryObjectives map[float64]float64
+
+	// OnRegisterError defines a method to call to when registering
+	// a metric with the registerer fails. Use nil to specify
+	// to panic by default when registering fails.
+	OnRegisterError func(err error)
+}
+
+// NewExporter returns a new prometheus exporter for prometheus metrics.
+func NewExporter(opts Options) (*Exporter, error) {
+	if opts.Registerer == nil {
+		opts.Registerer = prometheus.DefaultRegisterer
+	} else {
+		// A specific registerer was set, check if it's a registry and if
+		// no gatherer was set, then use that as the gatherer
+		if reg, ok := opts.Registerer.(*prometheus.Registry); ok && opts.Gatherer == nil {
+			opts.Gatherer = reg
+		}
+	}
+	if opts.Gatherer == nil {
+		opts.Gatherer = prometheus.DefaultGatherer
+	}
+	if opts.OnRegisterError == nil {
+		opts.OnRegisterError = func(err error) {
```

Comment:
Continuing the above note, I think we should try to propagate errors back to the `Meter`, which will have its own handler registered. See also #174.

Reply:
Ok, I added these as a way to deal with errors in the `Export` method. It looks like in the latest change `Export` can return errors, so then we shouldn't need this any more..

Reply:
Removed the `OnRegisterError` and propagated the errors.